### PR TITLE
SNOW-200637 support Arrow 2.0.x and ease building conda packages

### PR DESCRIPTION
**Pull request by xhochy**
The added environment variables would help us get rid of patching the package in the conda package.

### CODE REVIEW

**Comment by xhochy** [comment]
I guess tests are failing as this is a PR from a fork?

**Comment by sfc-gh-mkeller** [comment]
> I guess tests are failing as this is a PR from a fork?

Thanks @xhochy let me kick off tests for you!

**Comment by sfc-gh-mkeller** [comment]
I understand that the compiler flags would be helpful for you @xhochy 
While I figure out why the back-end doesn't like the new arrow version I'd happy to merge the compiler flags part of this PR if you split it into it's own PR.
I really appreciate your help!

**Comment by xhochy** [comment]
Anything that I could do to get this forward or even update to pyarrow 2.0?

**Comment by xhochy** [comment]
@sfc-gh-mkeller Updated to `pyarrow==2.0`, it works for me on my Snowflake instance when installed via `conda`, i.e. the server returns Arrow results for me there.

I'll try later using the wheel.

**Comment by xhochy** [comment]
Interestingly the wheel is failing with 

```
ImportError: /opt/conda/envs/sd-pricing-r/lib/python3.8/site-packages/snowflake/connector/arrow_iterator.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZNK5arrow16KeyValueMetadata5valueEl
```

Demangled name: `arrow::KeyValueMetadata::value(long) const`.

**Comment by xhochy** [comment]
> Interestingly the wheel is failing with
> 
> ```
> ImportError: /opt/conda/envs/sd-pricing-r/lib/python3.8/site-packages/snowflake/connector/arrow_iterator.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZNK5arrow16KeyValueMetadata5valueEl
> ```
> 
> Demangled name: `arrow::KeyValueMetadata::value(long) const`.

This is an issue only if you mix `pyarrow` from conda with the wheel. If you either fully use wheels from PyPI or packages from conda everything looks good.

**Comment by sfc-gh-mkeller** [comment]
> > Interestingly the wheel is failing with
> > ```
> > ImportError: /opt/conda/envs/sd-pricing-r/lib/python3.8/site-packages/snowflake/connector/arrow_iterator.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZNK5arrow16KeyValueMetadata5valueEl
> > ```
> > 
> > 
> > Demangled name: `arrow::KeyValueMetadata::value(long) const`.
> 
> This is an issue only if you mix `pyarrow` from conda with the wheel. If you either fully use wheels from PyPI or packages from conda everything looks good. So we're fine here.

This now conflicts with my work in  https://github.com/snowflakedb/snowflake-connector-python/pull/552

Let me try kicking off tests again and see how they do!

**Comment by xhochy** [comment]
> This now conflicts with my work in #552

I can rebase after #552 is merged, as CI is still failing in the tests, we probably still need some time to debug.

File: src/snowflake/connector/arrow_iterator.pyx [hunk 1]
```diff
@@ -88,6 +88,13 @@ cdef extern from "arrow/api.h" namespace "arrow" nogil:
     cdef cppclass CRecordBatchReader" arrow::RecordBatchReader":
         CStatus ReadNext(shared_ptr[CRecordBatch]* batch)
 
+    cdef cppclass CResult "arrow::Result"[T]:
+        CResult()
+        CResult(CStatus)
+        CResult(T)
+        c_bool ok()
+        CStatus status()
+        T operator*()
```
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/arrow_iterator.pyx:97]: Please look at line 81 and add the missing functions into the already existing definition.

**Comment by sfc-gh-mkeller** [comment]
@xhochy I just checked and you redefined `arrow::Result[T]`. I already added a definition for that in my PR.
Also let me reiterate that my other offer still stands for https://github.com/snowflakedb/snowflake-connector-python/pull/418#issuecomment-700909888
I'd be happy to merge a PR separately for the compiler flags you use on anaconda.

**Comment by xhochy** [comment]
> @xhochy I just checked and you redefined `arrow::Result[T]`. I already added a definition for that in my PR.
> Also let me reiterate that my other offer still stands for [#418 (comment)](https://github.com/snowflakedb/snowflake-connector-python/pull/418#issuecomment-700909888)

For me the value is in the combined patch, we can separate it out to get a smaller diff but the next time I would work on the build flags is when there is an update to the `pyarrow` dependency.

> I'd be happy to merge a PR separately for the compiler flags you use on anaconda.

It's conda-forge, not anaconda, they're different entities ;)

File: src/snowflake/connector/arrow_iterator.pyx (hunk 1 above)
Comment (xhochy) [reply src/snowflake/connector/arrow_iterator.pyx:97]: Out of curiosity: When aren't you using the offical `pxd` that is coming with `pyarrow`?

**Comment by sfc-gh-mkeller** [comment]
> It's conda-forge, not anaconda, they're different entities ;)

I apologize! Shows you how much I know about both of them 😓

File: src/snowflake/connector/arrow_iterator.pyx (hunk 1 above)
Comment (sfc-gh-mkeller) [reply src/snowflake/connector/arrow_iterator.pyx]: I didn't write this code and since I took over it I have been just following the same patterns.
To be honest it works and so I didn't see too much value in getting myself to know all the ins and outs of the Arrow code (when people are asking for so many other things to be fixed).

But if you could give me some pointers I'd be happy to look into it some more!

**Comment by xhochy** [comment]
Did some monkey-testing in #572 , this seems fine to retest again, @sfc-gh-mkeller can you push to your non-fork branch?

One thing I noticed it that Arrow tests are also run when pyarrow is not installed. This works as the package currently ships with the shared objects. I would be happy to provide a follow-up patch soon where we can exclude the `so` / `dylib`s and still be manylinux-compatible but I would do that in a separate PR ;) 

If things fail again, I would switch the Arrow-based tests to be only run when `pyarrow` is actually installed. It's kind of an edge case currently that part of the code works while the Python package itself is not available.

**Comment by sfc-gh-mkeller** [comment]
> One thing I noticed it that Arrow tests are also run when pyarrow is not installed. This works as the package currently ships with the shared objects. I would be happy to provide a follow-up patch soon where we can exclude the `so` / `dylib`s and still be manylinux-compatible but I would do that in a separate PR ;)

We do this on purpose, since Arrow is faster than json we wanted to make sure that most customer can use arrow result retrieval without installing `pyarrow`.

**Comment by sfc-gh-hyu** [comment]
> One thing I noticed it that Arrow tests are also run when pyarrow is not installed. This works as the package currently ships with the shared objects. I would be happy to provide a follow-up patch soon where we can exclude the `so` / `dylib`s and still be manylinux-compatible but I would do that in a separate PR ;)

@xhochy I tried to ship our python connector without so/dylib being bundled, but I failed :) Main reason is that I don't know how to dynamically linking to a pyarrow installed directory. Customer can have virtual environment anywhere on their box, which means pyarrow's so/dylib file is dynamic. Asking customer to manually set LD_LIBRARY_PATH is bad experience. It would be really helpful if you can somehow get rid of this bundling.

**Comment by sfc-gh-mkeller** [comment]
Merged through #419
